include/engine/guidance/leg_assembler.hpp

```diff
@@ -0,0 +1,160 @@
+        //
+        // The PathData will contain entries of b, c and d. But only c will contain
+        // a duration value since its the only point associated with a turn.
+        // As such we want to slice of the duration for (a,s) and add the duration for
+        // (c,d,t)
+        duration = duration - source_duration + target_duration;
+        auto summary = SummarizeRoute(route_data);
+
+        return RouteLeg{duration, distance, summary, boost::none};
+    }
+
+  private:
+    struct NamedSegment
+    {
+        double duration;
+        std::uint32_t position;
+        std::uint32_t name_id;
+    };
+
+    std::string SummarizeRoute(const std::vector<PathData> &route_data) const
+    {
+        // merges segments with same name id
+        const auto collapse_segments = [](std::vector<NamedSegment> &segments)
+        {
+            auto out = segments.begin();
+            auto end = segments.end();
+            for (auto in = segments.begin(); in != end; ++in)
+            {
+                if (in->name_id == out->name_id)
+                {
+                    out->duration += in->duration;
+                }
+                else
+                {
+                    ++out;
+                    BOOST_ASSERT(out != end);
+                    *out = *in;
+                }
+            }
+            return out;
+        };
+
+        std::vector<NamedSegment> segments(route_data.size());
+        std::uint32_t index = 0;
+        std::transform(route_data.begin(), route_data.end(), segments.begin(),
+                       [&index](const PathData &point)
+                       {
+                           return NamedSegment{point.duration_until_turn / 10.0, index++, point.name_id};
+                       });
+        // this makes sure that the segment with the lowest position comes first
+        std::sort(segments.begin(), segments.end(),
+                  [](const NamedSegment &lhs, const NamedSegment &rhs)
+                  {
+                      return lhs.name_id < rhs.name_id ||
+                             (lhs.name_id == rhs.name_id && lhs.position < rhs.position);
+                  });
+        auto new_end = collapse_segments(segments);
+        segments.resize(new_end - segments.begin());
+        // sort descending
+        std::sort(segments.begin(), segments.end(),
+                  [](const NamedSegment &lhs, const NamedSegment &rhs)
+                  {
+                      return lhs.duration > rhs.duration;
+                  });
+
+        // make sure the segments are sorted by position
+        const constexpr std::size_t MAX_USED_SEGMENTS = 2;
+        segments.resize(std::min(segments.size(), MAX_USED_SEGMENTS));
+        std::sort(segments.begin(), segments.end(),
+                  [](const NamedSegment &lhs, const NamedSegment &rhs)
+                  {
+                      return lhs.position < rhs.position;
+                  });
+
+        std::string summary;
```

Comment:
`boost::algorithm::join` does the next 10 lines; take a look at rtd/include/util/geojson

    join(segments | transformed(toName), ", ");